hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCleanerChore.java

```diff
@@ -0,0 +1,331 @@
+              LOG.warn("Directory {} was deleted during cleaner procedure execution,"
+                  + " skipping region {}",
+                storePath, regionPath.getName());
+              continue region;
+            }
+            RemoteIterator<LocatedFileStatus> rit = fs.listLocatedStatus(storePath);
+            List<Path> storeFiles = new ArrayList<Path>();
+            // Load list of store files first
+            while (rit.hasNext()) {
+              Path p = rit.next().getPath();
+              if (fs.isFile(p)) {
+                storeFiles.add(p);
+              }
+            }
+            LOG.info("Found {} store files in: {}", storeFiles.size(), storePath);
+            Path currentPath = null;
+            try {
+              for (Path pp : storeFiles) {
+                currentPath = pp;
+                LOG.trace("Store file: {}", pp);
+                HStoreFile sf =
+                    new HStoreFile(fs, pp, conf, CacheConfig.DISABLED, BloomType.NONE, true);
+                sf.initReader();
+                byte[] mobRefData = sf.getMetadataValue(HStoreFile.MOB_FILE_REFS);
+                byte[] bulkloadMarkerData = sf.getMetadataValue(HStoreFile.BULKLOAD_TASK_KEY);
+                // close store file to avoid memory leaks
+                sf.closeStoreFile(true);
+                if (mobRefData == null && bulkloadMarkerData == null) {
+                  LOG.warn("Found old store file with no MOB_FILE_REFS: {} - "
+                      + "can not proceed until all old files will be MOB-compacted.",
+                    pp);
+                  return;
+                } else if (mobRefData == null && bulkloadMarkerData != null) {
+                  LOG.debug("Skipping file without MOB references (bulkloaded file):{}", pp);
+                  continue;
+                }
+                // mobRefData will never be null here, but to make FindBugs happy
+                if (mobRefData != null && mobRefData.length > 1) {
+                  // if length = 1 means NULL, that there are no MOB references
+                  // in this store file, but the file was created by new MOB code
+                  String[] mobs = new String(mobRefData).split(",");
+                  if (LOG.isTraceEnabled()) {
+                    LOG.trace("Found: {} mob references: {}", mobs.length, Arrays.toString(mobs));
+                  } else {
+                    LOG.debug("Found: {} mob references", mobs.length);
+                  }
+                  regionMobs.addAll(Arrays.asList(mobs));
+                } else {
+                  LOG.debug("File {} does not have mob references", currentPath);
+                }
+              }
+            } catch (FileNotFoundException e) {
+              String warnMsg = String.format(
+                "Missing file:%s" + " Starting MOB cleaning cycle from the beginning due to error",
+                currentPath);
+              LOG.warn(warnMsg, e);
```

Comment:
Suitable method in org.slf4j.Logger?

Reply:
the slf4j Logger methods for messages take vararg to do matching to parameters and also have special handling so that the very last parameter can be an exception.

ref: http://www.slf4j.org/faq.html#paramException

Reply:
the slf4j Logger methods for messages take vararg to do matching to parameters and also have special handling so that the very last parameter can be an exception.

ref: http://www.slf4j.org/faq.html#paramException